R/util.R

```diff
@@ -130,27 +130,62 @@ util.is_vector_of_numbers <- function(x){
     # (regardless of whether x is numeric)
     # Character vectors sometimes need to be changed to numerics,
     # e.g., when all columns set to character types by default.
-
+  
+    # determine if x is logical first, because as.numeric will coerce logicals
+    # to boolean 0,1 rather than to NA, meaning util.is_vector_of_numbers would return TRUE
+    if(is.logical(x)) return(FALSE)
+    
     # find numeric values
-    numeric_values <- grepl("^-*[[:digit:]]*\\.*[[:digit:]]+$",x)
+    x_as_numeric <- suppressWarnings(as.numeric(x))
 
-    # find the blank values
-    blank_values <- util.is_blank(x)
+    # anything that gets coerced to NA by as.numeric is not really a number.
+    # but elements that were blank originally are ok.
+    blanks <- is.na(x_as_numeric)
+    coerced_blanks <- blanks & !util.is_blank(x)
 
-    # Return TRUE if all values are either numeric or blank
-    # Unless all values are blank, then false
-    # Return FALSE otherwise
-    if( all(blank_values) ){
+    # Return FALSE if there are any coerced blanks, because that means x 
+    # contained elements that could not be converted to numeric.
+    # Otherwise return TRUE.
+    if(any(coerced_blanks)){
         return(FALSE)
     }
-    else if(all(numeric_values | blank_values)){
+    else{
         return(TRUE)
     }
-    else{
-        return(FALSE)
+}
+
+
+util.is_vector_of_numbers_test <- function(){
+    x_char <- c("1", "a", "b", "1.1", "1e05")
+    x_number <- c("1", "1.1", "2", "-1.1")
+    x_scientific <- c("1e05")
+    x_with_blanks <- c(x_number, "", NA)
+    x_logical <- c(TRUE, FALSE)
+
+    if(util.is_vector_of_numbers(x_char) == TRUE){
+        stop("In util.is_vector_of_numbers, vectors containing " %+%
+                 "character values are being evaluted as numbers.")
+    }
```

Comment:
nit: I try to eliminate as many negatives and double negatives in code as possible, because humans are slow at reading them. The fact that the expression in the `if` condition is the opposite of what you're looking for ("stop if this is WRONG thing is true") threw me the first time. It's more normal, and requires fewer mental gymnastics, to make a positive statement like "I assert this is true". This is why, in my opinion, gymnast should define:

```
assert <- stopifnot
```

Then also note the error message of a failed `stopifnot` is actually pretty useful b/c it tells you the name of the argument you passed, so with your well-named inputs you can drop all the custom stop messages. Put it together, and the code becomes much more readable:

```
util.is_vector_of_numbers_test <- function(){
    x_char <- c("1", "a", "b", "1.1", "1e05")
    x_number <- c("1", "1.1", "2", "-1.1")
    x_scientific <- c("1e05")
    x_with_blanks <- c(x_number, "", NA)
    x_logical <- c(TRUE, FALSE)

    assert(!util.is_vector_of_numbers(x_char))
    assert(util.is_vector_of_numbers(x_number))
    assert(util.is_vector_of_numbers(x_scientific))
    assert(util.is_vector_of_numbers(x_with_blanks))
    assert(!util.is_vector_of_numbers(x_logical))
 }
 
 
util.is_vector_of_numbers_test()

# Assuming your fix wasn't in place:
# Error: util.is_vector_of_numbers(x_with_blanks) is not TRUE
```

Which I would read as something like "make sure the function correctly fails in this case" rather than "if this function fails to fail then stop."